Add logResources option

alutman-innablr opened the pull request:
https://github.com/Innablr/revolver/issues/142

* Changed the schema for `saveResources`. There's now a different field to control how resources are outputted at the start of execution.
* Added `console` and `csv` options to `resourceLog`
* Updated `audit.console` to use the logger rather than console.log
* TODO later, add `tags()` method to `toolingInterface` so the console and CSV can print tags.


Example
```
defaults:
  settings:
      resourceLog:
        json:
          file: resources.json
        console:
          reportTags: ["Name"]
        csv:
          file: resources.csv
          reportTags: ["Name"]
```

alutman-innablr committed:
Add logResources option, update readme for defaults

alutman-innablr committed:
Expanded resource log schema and options

simon-anz commented:
The names "audit" and "resourceLog" feel inconsistent.  Should the former be "auditLog"?

simon-anz commented:
Please include the account-name (from config), value of the availabilityTag (in that account, if overridden), and Name Tag.

simon-anz commented on lib/resourceLog.ts:
```diff
@@ -0,0 +1,60 @@
+  }
+  process(): void {
+    const header = `${'ACCOUNT_ID'.padEnd(16)}${'REGION'.padEnd(20)}${'TYPE'.padEnd(20)} ${'ID'.padEnd(40)} ${'STATE'}`;
+    const lines = this.entries.map((r) => `${(r.accountId || '').padEnd(16)}${(r.region || '').padEnd(20)}${(r.resourceType || '').padEnd(20)} ${r.resourceId.padEnd(40)} ${r.resourceState}`)
```
Can we do this (and the other similar lines) in a slightly more structured way, that's easier to view and edit.  Maybe a list of string+pad, then a loop/function that pads the values and applies the separators.

simon-anz replied:
Maybe

```typescript
    const lines = this.entries.map((r) => [
      (r.accountId || '').padEnd(16),
      (r.region || '').padEnd(20),
      (r.resourceType || '').padEnd(20),
      r.resourceId.padEnd(40),
      r.resourceState
    ].join(''));
```

Maybe it's not worth the bother??

simon-anz commented on lib/resourceLog.ts:
```diff
@@ -0,0 +1,60 @@
+  }
+  process(): void {
+    const header = `${'ACCOUNT_ID'.padEnd(16)}${'REGION'.padEnd(20)}${'TYPE'.padEnd(20)} ${'ID'.padEnd(40)} ${'STATE'}`;
+    const lines = this.entries.map((r) => `${(r.accountId || '').padEnd(16)}${(r.region || '').padEnd(20)}${(r.resourceType || '').padEnd(20)} ${r.resourceId.padEnd(40)} ${r.resourceState}`)
```
"CSV" is not C

alutman-innablr committed:
Make CSV actually csv

alutman-innablr committed:
Log account name

alutman-innablr committed:
Rename audit to auditLog

alutman-innablr committed:
Align auditLog and resourceLog in functionality

alutman-innablr commented:
> The names "audit" and "resourceLog" feel inconsistent. Should the former be "auditLog"?

Changed.

> Please include the account-name (from config), value of the availabilityTag (in that account, if overridden), and Name Tag.

Added the account name but the availability tag is a plugin setting. We'd either need a pretty hacky condition to look at all the plugins to see if there's one with an availabilityTag field or restructure how the resource log executes.
If it's important, I can add the hacky lookup.

simon-anz commented:
I think it's worthwhile adding this, as it is most-common use-case.  
You could push the responsibility for selecting "extra tags" up to the caller (AccountRevolver), and pass an `IncludeTags` parameter. The caller can look in powercycle+powercycleCentral (if enabled), and select their `.availabilityTag`

alutman-innablr committed:
Option to report tags in resource csv and console logs

alutman-innablr commented:
Added option to CSV and console to specify tags to report. Linking this value to the plugin config is a lot more pain (plugins can have several configs with different tags)

simon-anz reviewed (approved):
LGTM. Sample output:

<img width="1077" alt="image" src="https://github.com/Innablr/revolver/assets/69175652/f0c90b3c-4749-45cb-ac1f-e591d5a5c088">
